### ISSUE TITLE
UNIX ports fail to log their path when stopped

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4477]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4477|
|Type|defect|
|Created|2010-05-31 23:12:57Z|
|Branch|https://github.com/twisted/twisted/tree/port-connectionLost-logging-4477|

Consider the attached program.  When it is stopped, it logs these messages:

```
2010-05-31 19:09:54-0400 [twisted.internet.protocol.ServerFactory] (Port None Closed)
2010-05-31 19:09:54-0400 [twisted.internet.protocol.ServerFactory] (Port 42734 Closed)
```

The TCP message is nice.  It tells us which port being listened on was stopped (although it doesn't tell us it is TCP, that might be nice).  The UNIX message is not so nice.  It just says `None` where it should say some path.

Attachments:

* [noneport.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f0a/f0ac2325b6bd0302bcd8862ca5852f99dcfe5a91/5c1709f6f74a1d88c6d44a90e5c809234c5aebbc.py) (304 bytes) - added by exarkun on 2010-05-31 23:13:10Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4477 4477
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_port_connectionLost_logging_4477 branches/port-connectionLost-logging-4477
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1275347577000000 1275347577000000
changetime__1277860930000000 1277860930000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

(In [29271]) Branching to 'port-connectionLost-logging-4477'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli removed owner|
|-|-|

When noneport.py is run using the branch's twisted it produces:

```
2010-06-09 01:38:48-0400 [twisted.internet.protocol.ServerFactory] (UNIX Socket '/var/folders/KB/KB0si0qlEICYFN-C9OpNRE+++TI/-Tmp-/tmpk750_M' Closed)
2010-06-09 01:38:48-0400 [twisted.internet.protocol.ServerFactory] (TCP Port 52017 Closed)
```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

Oops, looks like "ref" doesn't work in place of "refs".  Please see r29272

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks.

  1. Both of the tests have the problem that they test the posixreactor-implementation-detail classes and hook objects up to the current global reactor.
  1. Instead they should be using `ReactorBuilder` to test the behavior of objects gotten back from reactor methods.  Otherwise, you also need tests for the iocpreactor-implementation-detail `Port` class (not as good).
  1. The docstring for `test_connectionLostLogMsg` is a little vague.  Just say what information the log message is supposed to convey - an address identifying the port and the fact that it was closed.
  1. `TestCase.addCleanup` takes *args, so you don't need a lambda, you can just pass the extra arguments.
  1. It's preferred to use tuples as the right hand argument to string %.  So `format % (repr(port.port),)`, for example.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [29352]) Changed the tests to use ReactorBuilder

refs [#4477](https://github.com/twisted/twisted/issues/4477)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [29353]) Changed the dosctring of the test to be more detailed.

refs [#4477](https://github.com/twisted/twisted/issues/4477)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli removed owner|
|-|-|


>   1. Both of the tests have the problem that they test the posixreactor-implementation-detail classes and hook objects up to the current global reactor.
>   1. Instead they should be using `ReactorBuilder` to test the behavior of objects gotten back from reactor methods.  Otherwise, you also need tests for the iocpreactor-implementation-detail `Port` class (not as good).

Done

>   1. The docstring for `test_connectionLostLogMsg` is a little vague.  Just say what information the log message is supposed to convey - an address identifying the port and the fact that it was closed.

Done

>   1. `TestCase.addCleanup` takes *args, so you don't need a lambda, you can just pass the extra arguments.

No longer applicabe

>   1. It's preferred to use tuples as the right hand argument to string %.  So `format % (repr(port.port),)`, for example.

Done

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks!

  1. The string interpolation still isn't using tuples as the right hand argument.  `("foo")` is still just a string, you need a comma to make a tuple: `("foo",)`.
    1. in `_logConnectionLostMsg` in twisted/internet/unix.py
    1. in `_logConnectionLostMsg` in twisted/internet/tcp.py
    1. in `getExpectedConnectionLostLogMsg` in twisted/internet/test/test_unix.py
    1. in `getExpectedConnectionLostLogMsg` in twisted/internet/test/test_tcp.py
  1. The docstring for `UNIXPortTestsBuilder` should talk about `IReactorUNIX.listenUNIX` instead of about `unix.Port`
  1. The docstring for `TCPPortTestsBuilder` should talk about `IReactorTCP.listenTCP` instead of about `tcp.Port`.
  1. `UNIXPortTestsBuilder` needs to skip if AF_UNIX sockets aren't supported.  The skip for the existing case (on `UNIXFamilyMixin` will probably work, although it looks like it's not strictly the correct skip - instead checking if the reactor provides IReactorUNIX would be better.
  1. The `getListeningPort` and `getExpectedConnectionLostLogMsg` methods should have docstrings.
  1. The unix socket being closed is a port too.  So I guess it'd be better to say _UNIX Port_ instead of _UNIX Socket_.  Although I'm not really sure about this.  So use whichever sounds best to you.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

(In [29407]) Changed string interpolation to use tuples, added docstrings for helper test methods, and skip the UNIX deprecation tests if the reactor does not provide IReactorUNIX.

refs [#4477](https://github.com/twisted/twisted/issues/4477)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @cyli|
|-|-|

  1. pyflakes points out one unused import in twisted/internet/tcp.py
  1. The two-argument form of raise is... I wish I could say "deprecated", but python-dev doesn't understand how to gracefully phase out old things, so it's not deprecated... _anachronistic_.  Always `raise Foo(bar)`, not `raise Foo, bar`
    1. If you want, you can also use the (mis/undocumented) `requiredInterface` attribute of `ReactorBuilder`.  Then the interface check will be done for you by `buildReactor`.
  1. Looks like IOCP reactor's message doesn't match the expected message:

```
twisted.trial.unittest.FailTest: '(TCP Port 3279 Closed)' not in ['(Port 3279 Closed)', '(Port 3279 Closed)', 'Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x03C29FD0>', 'Stopping factory <twisted.internet.protocol.ServerFactory instance at 0x03C29FD0>', 'Main loop terminated.']
```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

I have filed a ticket ([#4527](https://github.com/twisted/twisted/issues/4527)) regarding the invalid documentation for ReactorBuilder with regards to requiredInterface(s).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

(In [29451]) Added the new message to iocpreactor.tcp.Port as well

refs [#4477](https://github.com/twisted/twisted/issues/4477)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli removed owner|
|-|-|

Ok, removed the unused import, changed the raise to instead use the requiredInterface attribute, and changed t.i.iocpreactor.tcp to also log the "TCP Port" message (there seems to be a lot of the same code in t.i.iocpreactor.tcp as t.i.tcp).  The IOCPReactor part of the tests now pass.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/port-connectionLost-logging-4477

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli commented|
|-|-|

(In [29452]) Bump copyright for t.i.iocpreactor

refs [#4477](https://github.com/twisted/twisted/issues/4477)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @cyli|
|-|-|

> (there seems to be a lot of the same code in t.i.iocpreactor.tcp as t.i.tcp)

Indeed.  iocpreactor was implemented by copying most of the files in twisted/internet/ and then hacking on it. :/  I think there's a ticket somewhere to factor out the duplication.

Just one more thing:

```
twisted/internet/test/test_unix.py:27: 'unittest' imported but unused
```

But also remember not to trip yourself up over [#4527](https://github.com/twisted/twisted/issues/4527) and the `UNIXPortTestsBuilder` added in this branch. :)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli set status to `closed`|
|-|-|

(In [29480]) Merge port-connectionLost-logging-4477 : make port closing messages more specific

Author: cyli
Reviewer: exarkun
Fixes: [#4477](https://github.com/twisted/twisted/issues/4477)

When TCP and UNIX ports are closed, log a message about the type of port that was closed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
